### PR TITLE
Switch documentation to more modern theme

### PR DESCRIPTION
**Pull request by nilsvu**
## Proposed changes

This is a suggestion to switch our Doxygen theme to the one provided by https://github.com/jothepro/doxygen-awesome-css. We may want to switch to a full Doxygen+Sphinx+Breathe(+Exhale) stack at some point, but until that's working better with template code, I suggest to just make our documentation look better and work on mobile by adjusting the Doxygen theme.

To be clear, all this PR does is switch out the CSS files that determine how the Doxygen HTML displays in a browser, along with a few minor configuration things.

The site automatically displays in light mode depending on the browser configuration. It also works on mobile. See the repo linked above for details. Here's a screenshot in dark mode:

Bildschirmfoto 2021-08-04 um 13 26 15
Bildschirmfoto 2021-08-04 um 13 29 51

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `new feature` if appropriate.

### CODE REVIEW

File: docs/Doxyfile.in
Comment (wthrowe) [inline docs/Doxyfile.in:77]: Why drop the file pages?

File: docs/Contributing/Contributing.md
Comment (wthrowe) [inline docs/Contributing/Contributing.md:1]: Can we make the content wider somehow?  I get lots of blank space on the left and right and horizontal scrollbars for the stuff in the middle.

I'm not super enthusiastic about the tree on the left for similar reasons.  It's using valuable screen real estate that could go towards displaying content with less horizontal scrolling.

File: docs/DevGuide/DevGuide.md
Comment (wthrowe) [inline docs/DevGuide/DevGuide.md]: You've been a bit over-aggressive on the find-replace.

File: docs/Contributing/Contributing.md
Comment (wthrowe) [inline docs/Contributing/Contributing.md:2]: Is there a way to disable the animation on the left panel every time I click a link?

**Review by wthrowe — CHANGES_REQUESTED** [review]
Overall, this looks very nice.  I don't have any complaints about how the content looks, but I have some concerns about how the meta-information is displayed.

File: docs/DevGuide/DevGuide.md
Comment (nilsvu) [reply docs/DevGuide/DevGuide.md]: hehe good catch 😅

File: docs/Doxyfile.in
Comment (kidder) [reply docs/Doxyfile.in:77]: @wthrowe do you find the file pages useful?  Since we aren't documenting most files, I don't find it that useful.  If someone were to start exploring the documentation by browsing the file pages, they might get the impression we don't have much documentation...

**Comment by kidder** [comment]
How do I search?

File: docs/Contributing/Contributing.md
Comment (nilsvu) [reply docs/Contributing/Contributing.md:1]: Actually I think the ultra-wide layout of the old Doxygen theme is one of the main things that make it hard to read 😄 Lines should only be so long until they become hard to read. Look at pretty much any website these days: they have a max content width. GitHub has, for example.

Where are you getting horizontal scroll bars? Those are annoying, but I don't see any unless I make the window very narrow, but not narrow enough to get into mobile-friendly view.

File: docs/Doxyfile.in
Comment (nilsvu) [reply docs/Doxyfile.in:77]: Yea I figured it's better to save a little disk space and generation time instead of generating those file pages. I've never used them, but I can restore this if you find them useful 🤷‍♂️

Comment (nilsvu) [reply docs/Doxyfile.in:77]: I actually thought it would be nice to replace the Doxygen-generated file-links with links to the files on GitHub. That just needs a little Python-postprocessing of the HTML. It's not high-priority enough for me to do right now, but I thought I'd open an issue.

File: docs/Contributing/Contributing.md
Comment (nilsvu) [reply docs/Contributing/Contributing.md:2]: I haven't figured out yet when it appears. I think it's in https://github.com/doxygen/doxygen/blob/master/templates/html/navtree.js.

**Comment by nilsvu** [comment]
> How do I search?

There's a big search bar at the upper left 😄

File: docs/Contributing/Contributing.md
Comment (kidder) [reply docs/Contributing/Contributing.md:1]: I'm also not a big fan of the tree on the left.  So far I've looked with Safari and Chrome on my laptop;  I have not encountered any space on the right or horizontal scroll bars, but obviously haven't browsed everything...

**Comment by kidder** [comment]
The search bar is not showing up on my laptop for either browser...

**Comment by nilsvu** [comment]
@kidder ok that's concerning, I'll check. Did you download the docs from the CI artifact, or did you build it yourself (which Doxygen version)?

**Comment by kidder** [comment]
built it on my laptop.  doxygen --version  gives 1.8.16

File: docs/Contributing/Contributing.md
Comment (wthrowe) [reply docs/Contributing/Contributing.md:1]: Flowed text should not be wide, I agree, and the old style is bad with that.  But a lot of doxygen stuff can't be broken into multiple lines, and that stuff should be allowed to take as much space as it needs.

[GitHub won't let me attach an image, so here.](http://hosting.astro.cornell.edu/~wthrowe/horizontal_scrollbars.png)

File: docs/Doxyfile.in
Comment (wthrowe) [reply docs/Doxyfile.in:77]: I don't use them very often, so I'm not completely opposed to them going, but one thing I do use is the information about what files the documentation came from, and that seems to have gone with them.

File: docs/Contributing/Contributing.md
Comment (wthrowe) [reply docs/Contributing/Contributing.md:2]: The call to `docContent.animate` seems suspicious.

File: docs/Contributing/Contributing.md
Comment (kidder) [reply docs/Contributing/Contributing.md:1]: Ah okay, I do have a horizontal scroll-bar, however it is invisible until I scroll...

File: docs/Doxyfile.in
Comment (kidder) [reply docs/Doxyfile.in:77]: ah yes, that information is useful, so I would not like to see it disappear

**Review by geoffrey4444 — CHANGES_REQUESTED** [review]
I built the docs using this branch and put them here: http://www.tapir.caltech.edu/~geoffrey/spectre/docs/html/index.html

Is there a way to disable the sidebar? Otherwise it's useless on mobile.

**Comment by nilsvu** [comment]
@geoffrey4444 That doesn't currently look like it does on my machine. Can you try building it with Doxygen 1.9.1?

**Comment by geoffrey4444** [comment]
Works great on mobile using doxygen 1.9.1 (newest available via brew). Previously, I was using 1.8.5, which is the latest available on spack.

I updated http://www.tapir.caltech.edu/~geoffrey/spectre/docs/html/index.html with the 1.9.1 build, seems good

**Comment by nilsvu** [comment]
@geoffrey4444 Thanks for checking! Spack has doxygen@1.9.1 since March, I think you just need to `git pull` in the Spack directory :)

**Comment by nilsdeppe** [comment]
I think this looks really great! Thanks for working on this :) Some thoughts

 - Can't find the `namespaces` list. I use this quite a bit. I eventually found this under `Code reference`. I wonder if we should make the `Topics`, `Namespace List`, `Class List`, and `Class Index` be in the main group since this is what a lot of people frequently want access to.
 - Is it possible to do wrapping in the side menu? Some things have long "short names" and it would be useful to show the whole name, but I agree that we don't want the side bar becoming huge.
 - Regarding the wide view, I have no preference, only a wish that Doxygen could clang format the function signature and show that :D
 - Should we print a CMake warning if an older Doxygen version is detected that will produce sub-optimal output?

File: docs/Doxyfile.in
Comment (nilsvu) [reply docs/Doxyfile.in:77]: 👍 added this info back (`SHOW_USED_FILES = YES`)

**Comment by nilsvu** [comment]
The container has Doxygen 1.9.1 now, I'll push it once this is ready to merge.

- Added warnings when the Doxygen version may be incompatible. As you currently see in this PR, the CI test also fails with the warning. It will go away once I push the new container.
- I have not tried to do word-wrapping in the sidebar, that seems like an upstream issue. There's a bunch of things that would improve the sidebar, like more visible grouping and/or auto-expand the sub-items. That would also solve the visibility issue of the "Code reference" sub-items. Either needs upstream contributions (Doxygen and/or https://github.com/jothepro/doxygen-awesome-css) or some HTML/CSS/JS work on our side.

**Comment by nilsdeppe** [comment]
Looks like something's still going on with the versioning in CI

**Comment by nilsvu** [comment]
@nilsdeppe How so? CI fails because I haven't pushed the container yet (see comment above).

**Comment by nilsdeppe** [comment]
ah, I thought you pushed the container as soon as the file was container file was updated, which is what we've always done in the past. sorry

**Comment by nilsvu** [comment]
Pushed it now.

**Review by nilsdeppe — COMMENTED** [review]
Okay, I'm happy to merge this as-is. Maybe we should have an issue to track the concerns people raised, even if they end up being resolved upstream. I'll download the archived dox once CI completes just to double check things are functioning as expected.

**Comment by wthrowe** [comment]
I'm not OK with pushing everyone's concerns to a bug unless there is a plan to actually address that bug.

**Comment by nilsdeppe** [comment]
I don't know which of the concerns are left unaddressed. Could you mark discussions that have been resolved as "Resolved"? I also don't know which of the comments you view as blockers vs "This would be nice". Could you clarify which ones you require to be resolved so that any efforts can be concentrated on those?

**Comment by wthrowe** [comment]
I've marked the things I can, but most of the discussion on this PR is top-level comments so I can't mark them resolved or not.

In terms of blocking issues:
* We have one report that search doesn't work.
* I still think the sidebar makes the page very hard to use.  It takes up over a quarter of the screen for me.  Functions have to scroll sideways.  Long equations are just truncated with no way to scroll.

And, quite frankly, you know as well as I do that the issue tracker is where problems go to die.  Unless it's a compilation or test failure or there is an immediate plan to fix it, it's going to be ignored indefinitely.  I just looked, and Jun-Aug we closed 4 issues outside of those categories.  In that time there were 27 issues opened that are still open.

**Comment by nilsdeppe** [comment]
okay, thanks for spelling those out :) I agree that issues tend not to get fixed unless they are actually interfering directly with something. I'll let @nilsleiffischer comment on the feasibility of fixing those issues. I'd guess the side bar thing might be what prevents us from merging this. 

Related, what are your thoughts on the ReadTheDocs format? I think the standard style has a side bar, eg https://charm.readthedocs.io/en/latest/ It would be good to know now if this style would block any efforts to revamp the documentation before more effort is put into Sphinx+Breathe+ReadTheDocs.

**Comment by wthrowe** [comment]
Hard to judge, since Charm docs seem to be entirely prose, which looks fine but is also fine in this PR.  It is nice that nothing seems to be animated, but that's in the "annoying but not blocker" category anyway.

**Comment by nilsdeppe** [comment]
Okay, I certainly agree with your complaints about the format. I don't love the current Doxygen either, but we certainly shouldn't change unless everyone agrees it's a net improvement :) I'm still hoping that Breathe will improve to the point where it's useful. I think currently it takes over 10 minutes to run on our code base, mostly because it uses an XML parser that's super slow and has huge memory leaks (sigh...)

**Comment by nilsvu** [comment]
- The search issue was a Doxygen versioning thing. It works in the container, and with Doxygen versions 1.9.1 and above (maybe also older versions, I haven't tested any). Both CMake and `make doc(-check)` now print a warning when the version may not be compatible. Since Doxygen can (and does) change their HTML/CSS/JS with every version, we can really only reliably support a particular Doxygen version with our CSS/JS/postprocessing.
- @wthrowe we seem to have different priorities concerning the documentation layout. For me, it's absolute highest priority that the docs don't look like a website from the 90s, that they work on mobile, and that the prose guides and docstrings are well-readable. To accomplish the latter, I prefer the max-content-width over the unconstrained width by a lot, even if that leads to some horizontal scrollbars in very long function declarations. Currently I actually often find myself resize the browser window so the docstrings have a reasonable width so I can read them better. I find it much more important to have the function _documentation_ well-readable than the _declaration_. I actually fixed an upstream bug that made the function declarations render too large, so they should be a bit better now. Also, I'm willing to compromise -- it's very easy to adjust the content width from a technical standpoint, so if you find a width of 1000px (current setting) too narrow we can increase it a bit (you can try it yourself by adjusting the `--content-maxwidth` CSS variable in your browser. Set it to `auto` to try an unconstrained width. Try different browser window sizes, because people use a large variety of screens.).
- I personally like the sidebar, because it accomplishes what a site navigation must: Show you where you are, and where you can go. It's particularly useful to show you the available prose guides (dev guides and tutorials). I find the increased discoverability of those very valuable, compared to what we currently have. I also find the sidebar useful when browsing code docs, because it points to related functions and classes. @wthrowe I do share your dislike of the animation, so I removed it, along with some other modifications. @nilsdeppe I also agree it’s better to show the „Code reference“ subitems always, so I do that now.

I pushed an updated version, so do have a look when you get the chance. You can download the rendered docs from the "Checks > Artifacts" in this PR once CI is done.

**Comment by nilsdeppe** [comment]
Rendering this locally with `tools/pr-docs` I don't seem to have a search bar. Navigation seems to be working fine otherwise, and I can now find the namespace list, etc.

Not necessarily in this PR, but should we move `Code Reference->Topics` to a main heading named something like `Code Modules`?

Another thing in a followup: should we make the side bar skinnier? If I use half a 13" laptop screen then the sidebar uses near 50% of the window real estate. Reducing that to say 33% might feasible :shrug:

**Comment by nilsvu** [comment]
- `pr-docs` works for me. Could you make sure you're using a recent Doxygen version? Doxygen 1.9.1 was released in January, that should work. `pr-docs` should have printed a warning as well.
- Fine with me, and done.
- Reduced the sidebar width.

**Review by wthrowe — DISMISSED** [review]
My priority is that as much of the important information as possible is visible on the screen in a reasonable-sized browser window.  For some of our pages it's not possible for that to be all of it, because of wide content like signatures and equations, but it would be good if as many places worked as possible.  (In case I haven't made it clear, I've generally been talking about "narrow" browser windows, like those covering only half a widescreen monitor.)  I think that this change is sacrificing functionality for aesthetics.

Anyway, I won't block this anymore, since I don't usually use the web interface so I'm not the target audience.

On a few particular points:

Shrinking the sidebar helped quite a bit.  I would go further, since it is still larger than any other site that I regularly visit.  (It's also still animated, but that's a minor annoyance.)

I agree that the sidebar is nice for the guides, but for code I just don't find being shown an apparently unsorted list of every class in (one of) the groups of the thing I'm looking at to be worth the screen real estate.

It is still bad that long equations are truncated, but we don't have too many of those.

Search bar is working fine for me with 1.9.1.

**Comment by nilsdeppe** [comment]
Ah, @wthrowe are you referring to equations like the ABN ones? It seems like in the current develop dox if I make the window narrower a horizontal scrollbar is added while with this PR it isn't. I agree it would be nice not to have such a functionality regression. I don't know enough about CSS/HTML/web things to know why there's a difference or how it could be fixed. I do worry about this because we have documentation requests for things like how we do DG in the evolution code that will have a lot of equations, some of which will almost certainly be longer.

- I agree, thinner sidebar is good, could go a bit thinner still. Currently only "Performance, Efficiency, and Optimization" is cutoff. Maybe there's a way to have the contents wrap?
- That's reasonable. I don't understand what this little blue double arrow button should do. It made me think of having a button to hide the side bar, but that seems like a bigger endeavor...
- Search bar seems to work fine for me now too :shrug:

**Comment by nilsvu** [comment]
- Made the sidebar a bit thinner. Note that I chose a width of 300px before, to match ReadTheDox. Now, it's thinner than RTD.
- Removed that small blue button. It's generated by Doxygen and supposed to enable/disable "nav sync", but I found it confusing as well and not useful, so it's gone now.
- Some long equations show horizontal scrollbars for me (e.g. in the XCTS namespace docs) and others don't (e.g. in the ABN `compute_boundary_delta` function docs). My take on this is that we should try to keep equations to a reasonable length anyway, so when we encounter long equations that don't display nicely we should add some reasonable line breaks.
- I'm reaching a limit on how much time I can spend on this PR. In my opinion this improves the documentation significantly, in particular the discoverability and readability of dev guides. I don't see the Breathe+Sphinx pipeline working any time soon, so I think this is our best bet of modernizing the docs with reasonable effort. There _will_ be things that don't display perfectly (like long equations), and we can fix those when we encounter them. I have spent some time coming up with this configuration and find it works sufficiently well, but I don't have the time to do any major webdesign work right now.

**Review by nilsdeppe — DISMISSED** [review]
Okay, I'm happy with this.

@geoffrey4444 you currently have changes requested, can you please look at this?

**Review by geoffrey4444 — APPROVED** [review]
I don't need to see this again...everything looks good, except the dev guide page is broken.  No entries appear after `Getting started with improving`. Maybe a typo in that entry? I suggest fixing that as part of this PR or as a quick followup

http://www.tapir.caltech.edu/~geoffrey/spectre/docs/html/dev_guide.html

**Comment by nilsvu** [comment]
@geoffrey4444 could you double-check that you built the docs on the latest version of this branch? I remember seeing that bug a while ago and fixed it.

**Comment by geoffrey4444** [comment]
@nilsleiffischer I checked, and I must have had an old version, looks good!

**Review by geoffrey4444 — DISMISSED** [review]

**Comment by nilsvu** [comment]
Rebased on #3522

**Review by nilsdeppe — DISMISSED** [review]

**Comment by nilsdeppe** [comment]
Does this fix https://github.com/sxs-collaboration/spectre/issues/1887 ?

@wthrowe @kidder could you please let us know if you're happy with this or if you want changes?

**Comment by nilsvu** [comment]
@nilsdeppe #1887 is fixed already by upgrading Doxygen in the container

**Comment by nilsdeppe** [comment]
@wthrowe @kidder could you please take a look to see if you're happy with this?

**Review by nilsdeppe — APPROVED** [review]

**Review by kidder — APPROVED** [review]

**Pull request merged**
